sdk/storage/Azure.Storage.Files.DataLake/tests/DirectoryClientTests.cs

```diff
@@ -935,6 +937,164 @@ public async Task SetAccessControlAsync_ConditionsFail()
+            DataLakeDirectoryClient subdirectory1 = await directory.CreateSubDirectoryAsync(GetNewDirectoryName());
+            DataLakeFileClient file1 = await subdirectory1.CreateFileAsync(GetNewFileName());
+            DataLakeFileClient file2 = await subdirectory1.CreateFileAsync(GetNewFileName());
+            DataLakeDirectoryClient subdirectory2 = await directory.CreateSubDirectoryAsync(GetNewDirectoryName());
+            DataLakeFileClient file3 = await subdirectory2.CreateFileAsync(GetNewFileName());
+            DataLakeFileClient file4 = await directory.CreateFileAsync(GetNewFileName());
+
+            // Act
+            ChangeAccessControlListResult result = await directory.ModifyAccessControlListRecursiveAsync(
+                accessControlList: AccessControlList,
+                batchSize: 2);
+
+            // Assert
+            Assert.AreEqual(3, result.DirectoriesSuccessfulCount);
+            Assert.AreEqual(4, result.FilesSuccessfulCount);
+            Assert.AreEqual(0, result.FailureCount);
+        }
+
+        [Test]
+        [ServiceVersion(Min = BlobClientOptions.ServiceVersion.V2019_12_12)]
+        public async Task ModifyAccessControlRecursiveAsync_InBatches_WithProgressMonitoring()
+        {
+            await using DisposingFileSystem test = await GetNewFileSystem();
+            DataLakeDirectoryClient directory = await test.FileSystem.CreateDirectoryAsync(GetNewDirectoryName());
+            DataLakeDirectoryClient subdirectory1 = await directory.CreateSubDirectoryAsync(GetNewDirectoryName());
+            DataLakeFileClient file1 = await subdirectory1.CreateFileAsync(GetNewFileName());
+            DataLakeFileClient file2 = await subdirectory1.CreateFileAsync(GetNewFileName());
+            DataLakeDirectoryClient subdirectory2 = await directory.CreateSubDirectoryAsync(GetNewDirectoryName());
+            DataLakeFileClient file3 = await subdirectory2.CreateFileAsync(GetNewFileName());
+            DataLakeFileClient file4 = await directory.CreateFileAsync(GetNewFileName());
+
+            Mock<IProgress<ChangeAccessControlListPartialResult>> progresMonitorMock = new Mock<IProgress<ChangeAccessControlListPartialResult>>();
+
+            int batchSize = 2;
+
+            // Act
+            ChangeAccessControlListResult result = await directory.ModifyAccessControlListRecursiveAsync(
+                accessControlList: AccessControlList,
+                batchSize: batchSize,
+                progressHandler: progresMonitorMock.Object);
+
+            // Assert
+            Assert.AreEqual(3, result.DirectoriesSuccessfulCount);
+            Assert.AreEqual(4, result.FilesSuccessfulCount);
+            Assert.AreEqual(0, result.FailureCount);
+
+            progresMonitorMock.Verify(
+                x => x.Report(It.Is<ChangeAccessControlListPartialResult>(arg => arg.DirectoriesSuccessfulCount + arg.FilesSuccessfulCount <= batchSize)),
+                Times.Exactly(4));
+        }
+
```

Comment:
Are the following 2 P0 scenarios covered in testing? See [doc](https://microsoft.sharepoint.com/:w:/r/teams/AzureStorage/_layouts/15/doc2.aspx?sourcedoc=%7bb17c341f-d728-49fb-8848-043c66f088fb%7d&action=edit&wdPid=640e41f3&CID=D695C3BF-00F7-4FDE-BE02-ED79B1E9828A&wdLOR=cC08D1C9D-9292-49B3-8C54-3C2E98A120F6). Powershell will follow that in Powershell implematation and testing.
1.
> Track and log failures during recursive ACL application process, along with information required to reprocess failed directories/files

2.
> Provide user the choice between following options, in the event of the first failure: 
> Terminate recursive ACL application process and restart process from the beginning 
> Continue recursive ACL application process. Log information on failed batches/files, that will help reprocess these failures. Notify user after completion of process, to reprocess failed batches/files. (if this option is chosen it will apply for the first and all subsequent failures for the process)

Reply:
SetAccessControlRecursiveAsync_InBatches_WithProgressMonitoring should give you a good idea how to do the setup for those 2 requirements. `IProgress<ChangeAccessControlListPartialResult>` callback gives both progress and failed entries.

We don't have a test case that actually consumes failed entries yet as we were not yet able to set up account in such a way that we'd get any error.

Reply:
@kasobol-msft ，
Per the discussion with Ashish and Priya last week, Ashish will contact feature team to see how to cover the failure scenario test. 
We should not miss the test of them, as they are P0 scenario. We can't say our code definitely works well without test.

Besides that, I also don't find how to configure failure handler in case `SetAccessControlRecursiveAsync_InBatches_WithProgressMonitoring`, could you give some hint like which line of code will handle the failures?

Reply:
@blueww I addedd Set/Modify/RemoveAccessControlRecursiveAsync_WithProgressMonitoring_WithFailure test cases. They have one permission failure in the result. 
The idea is to implement IProgress interface and pass it as input while calling ACL recursive methods. IProgress is called back with counters from current batch and list of failed paths.

Reply:
@kasobol-msft 
Would you please point out which file has the failure? 
Seems I don't find them in the PR.

Reply:
Please see here on branch https://github.com/Azure/azure-sdk-for-net/blob/4e336bfdaf8b9e008b39532ec781ea6678ef208a/sdk/storage/Azure.Storage.Files.DataLake/tests/DirectoryClientTests.cs#L1021-L1060